src/ILCompiler.ReadyToRun/src/Compiler/DependencyAnalysis/ReadyToRunCodegenNodeFactory.cs

```diff
@@ -295,6 +295,28 @@ public IMethodNode CreateUnboxingStubNode(MethodDesc method, mdToken token)
             return signature;
         }
 
+        private readonly Dictionary<ReadyToRunFixupKind, Dictionary<TypeDesc, TypeFixupSignature>> _typeSignatures =
+            new Dictionary<ReadyToRunFixupKind, Dictionary<TypeDesc, TypeFixupSignature>>();
+
+        public TypeFixupSignature TypeSignature(ReadyToRunFixupKind fixupKind, TypeDesc typeDesc, SignatureContext signatureContext)
+        {
+            Dictionary<TypeDesc, TypeFixupSignature> perFixupKindMap;
+            if (!_typeSignatures.TryGetValue(fixupKind, out perFixupKindMap))
+            {
+                perFixupKindMap = new Dictionary<TypeDesc, TypeFixupSignature>();
+                _typeSignatures.Add(fixupKind, perFixupKindMap);
+            }
+
+            TypeFixupSignature signature;
+            if (!perFixupKindMap.TryGetValue(typeDesc, out signature))
+            {
+                EETypeNode.CheckCanGenerateEEType(this, typeDesc);
```

Comment:
Do we still need the `CheckCanGenerateEEType` calls in `ExternalTypeNode` and `AvailableType`? This call is pretty expensive.

Reply:
Hmm. EETypeNode and ExternEETypeSymbolNode do have the check - I guess that's what Simon used as the primary inspiration for AvailableType / ExternalTypeNode. I mean - I can easily remove the check from AvailableType / ExternalTypeNode but I'm in doubt whether it should be also removed from EETypeNode and ExternEETypeSymbolNode. After all, these instance also get constructed as part of dependency analysis, not compilation (by virtue of being returned from GetStaticDependencies / GetNonRelocationBasedDependencies) so the checks in their constructors also have the potential to tear down the AOT compiler and the only way to prevent it is to employ an upfront check - so that the check in the constructors can be either removed completely or changed into something like an assert (which would require a try block as the function unfortunately reports failure by throwing an exception).

Reply:
EETypeNode/ExternalTypeNode are the only places that do the check on the AOT side and we do make sure that whenever we see a new type, it goes through there and there's a try/catch guarding the exception. The exception that might be thrown from there gets handled (with the "generate throwing body" code). The check is in the right spot.

AvailableType/ExternalType nodes on the other hand seem to be only called from ComputeNonRelocationBasedDependencies routines of other nodes. If it ends up throwing, the exception is not recoverable and will tear down the compiler. Such behavior would be fine in debug builds, so maybe putting it in `#if DEBUG` would be fine too. I don't think it's useful for retail builds and it only slows down the build.

Reply:
The intent behind putting `CheckCanGenerateEEType` in `ExternalTypeNode` and `AvailableType` was to let us catch types giving us problems at table emission time. We have solved that a different way with ready-to-run by throwing `RequiresRuntimeJitException` so these calls to `CheckCanGenerateEEType` may be adding zero and something we can live without.